sequence_task_1/student/include/impl.h

```diff
@@ -1,51 +1,131 @@
-#pragma once
+#ifndef _SEQUANCE_TASK_1_
+#define _SEQUANCE_TASK_1_
+
 #include <string>
+#include <memory>
+#include <utility>
 
 struct Song
 {
     std::string name;
     explicit Song(std::string song_name)
-        : name {std::move(song_name)}
+        : name{std::move(song_name)}
     {
     }
 };
 
-bool operator==(const Song& lhs, const Song& rhs)
+bool operator==(const Song &lhs, const Song &rhs)
 {
     return lhs.name == rhs.name;
 }
 
-template<class Container>
+template <class Container>
 class StaticPlaylist
 {
 public:
     /** @todo Member traits */
 
+    typedef typename Container::size_type           size_type;
+    typedef typename Container::value_type          value_type;
+    typedef typename Container::reference           reference;
+    typedef typename Container::const_reference     const_reference ;
+    //typedef typename Container::type type;
+
     /** @todo Iterators */
+    typedef typename Container::iterator            iterator;
+    typedef typename Container::const_iterator      const_iterator;
+    typedef typename iterator::difference_type      difference_type;
+    typedef typename iterator::pointer              pointer;
+    typedef const pointer                           const_pointer;
+
+    const_iterator begin()
+    {
+        return m_tracklist.begin();
+    }
+
+    const_iterator end()
+    {
+        return std::end(m_tracklist);
```

Comment:
prefer one style of getting begin/end iterators
in your case you don't need general "std::end" call